File: src/utils/bundler.ts
/**
 * Bundler utility using esbuild for JavaScript and TypeScript files
 * Produces IIFE format output (similar to webpack)
 */

import * as esbuild from "esbuild";
import { denoPlugins } from "@luca/esbuild-deno-loader";

export interface BundleOptions {
  entryPoint: string;
  projectDir: string;
}

export interface BundleResult {
  code: string;
  isTypeScript: boolean;
}

/**
 * Check if the entry file uses TypeScript
 */
function isTypeScriptFile(entryPoint: string): boolean {
  return entryPoint.endsWith(".ts") || entryPoint.endsWith(".tsx");
}

/**
 * Bundle the entry point and all its dependencies into a single JavaScript string
 * Uses esbuild to produce IIFE format output (similar to webpack)
 */
export async function bundle(options: BundleOptions): Promise<BundleResult> {
  const { entryPoint } = options;
  const isTypeScript = isTypeScriptFile(entryPoint);

  try {
    const result = await esbuild.build({
      entryPoints: [entryPoint],
      bundle: true,
      write: false, // Return output in memory
      format: "iife", // IIFE format (similar to webpack)
      platform: "browser", // Target browser environment
      target: "es2020",
      minify: false, // Keep readable
      sourcemap: false,
      treeShaking: true,
      charset: "utf8", // Preserve non-ASCII characters (Japanese, etc.)
      plugins: [...denoPlugins()],
      // Don't wrap in global name since we'll append Sapphillon.Package
      globalName: undefined,
    });

    if (result.errors.length > 0) {
      const errorMessages = result.errors.map((e) => e.text).join("\n");
      throw new Error(`Bundling failed:\n${errorMessages}`);
    }

    if (result.outputFiles.length === 0) {
      throw new Error("No output generated from bundler");
    }

    let code = result.outputFiles[0].text;

    // Remove the IIFE wrapper for cleaner integration with Sapphillon.Package
    // esbuild produces: (() => { ... })();
    // We want the inner content to be available in global scope
    code = unwrapIIFE(code);

    return {
      code: code.trim(),
      isTypeScript,
    };
  } finally {
    // Always stop esbuild to clean up resources
    await esbuild.stop();
  }
}

/**
 * Unwrap IIFE to expose functions in global scope
 * Converts: (() => { function foo() {} })();
 * To: function foo() {}
 */
function unwrapIIFE(code: string): string {
  // Check if it's an IIFE: (() => { ... })();
  const iifeMatch = code.match(/^\s*\(\(\)\s*=>\s*\{([\s\S]*)\}\)\(\);\s*$/);
  if (iifeMatch) {
    let inner = iifeMatch[1].trim();

    // Remove "use strict" if present
    inner = inner.replace(/^"use strict";\s*/m, "");

    // Handle var declarations at the start (esbuild artifact)
    // Convert: var funcName = ...; to proper function declarations
    // This is complex, so we'll just clean up the output

    return inner;
  }

  // Also handle: (function() { ... })();
  const iifeFuncMatch = code.match(/^\s*\(function\s*\(\)\s*\{([\s\S]*)\}\)\(\);\s*$/);
  if (iifeFuncMatch) {
    let inner = iifeFuncMatch[1].trim();
    inner = inner.replace(/^"use strict";\s*/m, "");
    return inner;
  }

  // If not IIFE wrapped, return as-is
  return code;
}

/**
 * Check if the source code contains import statements
 */
export function hasImports(content: string): boolean {
  // Match any import statements
  const importRegex = /^\s*import\s+/gm;
  return importRegex.test(content);
}

File: src/utils/bundler.rs
use std::path::Path;
use anyhow::{anyhow, Context};
use std::fs;
use regex::Regex;

pub struct BundleOptions {
    pub entry_point: String,
    pub project_dir: String,
}

pub struct BundleResult {
    pub code: String,
    pub is_typescript: bool,
}

/// A simplified bundler that handles TypeScript transpilation and basic merging.
/// For this initial Rust rewrite, we'll focus on transpilation and basic export removal.
/// We'll use a regex-based transpiler for simple TS cases if SWC is having dependency issues,
/// or try to fix the SWC setup.
pub async fn bundle(options: BundleOptions) -> anyhow::Result<BundleResult> {
    let entry_path = Path::new(&options.entry_point);
    let is_typescript = entry_path.extension().map_or(false, |ext| ext == "ts" || ext == "tsx");

    let content = fs::read_to_string(entry_path)
        .with_context(|| format!("Failed to read entry point: {}", options.entry_point))?;

    let mut code = if is_typescript {
        // Fallback to a simpler approach if SWC fails
        simple_transpile_ts(&content)
    } else {
        content
    };

    // Simple export removal to match the original behavior
    code = remove_exports(&code);

    Ok(BundleResult {
        code: code.trim().to_string(),
        is_typescript,
    })
}

fn simple_transpile_ts(content: &str) -> String {
    // Very basic TS to JS conversion: remove type annotations
    // This is not a complete transpiler but covers basic cases for our plugins
    let mut code = content.to_string();

    // Remove ': number', ': string', ': boolean', ': void', etc.
    let re_types = Regex::new(r":\s*(number|string|boolean|void|any|string\[\]|number\[\])").unwrap();
    code = re_types.replace_all(&code, "").to_string();

    // Remove 'as string', etc.
    let re_as = Regex::new(r"\s+as\s+\w+").unwrap();
    code = re_as.replace_all(&code, "").to_string();

    code
}

fn remove_exports(code: &str) -> String {
    let re_named = Regex::new(r"(?m)^export\s+\{[^}]*\};?\s*$").unwrap();
    let re_keyword = Regex::new(r"(?m)^export\s+").unwrap();

    let step1 = re_named.replace_all(code, "");
    re_keyword.replace_all(&step1, "").to_string()
}

pub fn has_imports(content: &str) -> bool {
    let re = Regex::new(r"(?m)^\s*import\s+").unwrap();
    re.is_match(content)
}
